# Ebay Car Data Analysis

This project will clean and analyze data regarding cars for sale on ebay. The data was web scraped, and uploaded to Kaggle. The columns are the following:

**Columns:**              
 - dateCrawled         
 - name                
 - seller              
 - offerType           
 - price               
 - abtest              
 - vehicleType         
 - yearOfRegistration  
 - gearbox             
 - powerPS             
 - model               
 - kilometer           
 - monthOfRegistration 
 - fuelType            
 - brand               
 - notRepairedDamage   
 - dateCreated         
 - nrOfPictures        
 - postalCode          
 - lastSeenOnline       

In [36]:
import numpy as np
import pandas as pd

autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [37]:
# print the first and last few rows, as well as info about the data frame
autos
print(autos.head())
autos.info()

           dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere 

# Observations:

As you can see from the data output above, it appears that 'private' in the 'seller' column has been mispelled to 'privat'. Also, the vehicleType, gearbox, model, fuelType, and notRepairedDamage have null values. The price and odometer fields are text fields.

In [38]:
# print the existing column names:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [39]:
# rename the columns
new_cols = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_pictures', 'postal_code',
       'last_seen']

#assign modified column names back to columns attribute of data frame
autos.columns = new_cols

In [40]:
#view the data frame's first few rows
print(autos.head())

          date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158 

# Column name updates

The columns have been renamed to the following, for updated notation and clarity:

 - registration_year
 - registration_month
 - unrepaired_damage
 - ad_created
 - date_crawled
 - offer_type
 - vehicle_type
 - power_ps
 - fuel_type
 - num_pictures
 - postal_code
 - last_seen_online

In [41]:
# describe the data frame
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-23 18:39:34  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_pictures   postal_code            last_seen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

# Data set findings

'privat' in the seller columns is mispelled. Registration year appears to be a float data type. 'manuell' is mispelled in the gearbox column. odometer has 'km' and does not appear to be a text field. Registration month appears to be a float data type. The 'price' field has a $ in the values, so it would have to be a text type.

In [42]:
# replace price text values to be float data type
new_prices = []
for value in autos['price']:
    float_val = value.replace('$','')
    float_val = float_val.replace(',','')
    float_val = float(float_val)
    new_prices.append(float_val)
    
autos['price'] = new_prices

In [43]:
# replace odometer text values to be float data type
new_odm = []
for record in autos['odometer']:
    float_value = record.replace('km','')
    float_value = float_value.replace(',','')
    float_value = float(float_value)
    new_odm.append(float_value)
    
autos['odometer'] = new_odm

In [68]:
# rename 'odometer' column to 'odometer_km'
autos.rename(columns={"odometer":"odometer_km"}, inplace=True)
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [69]:
# view sorted frequency count of unique values for the odometer_km column
autos['odometer_km'].value_counts().sort_index(ascending=True).head(50)

5000.0        967
10000.0       264
20000.0       784
30000.0       789
40000.0       819
50000.0      1027
60000.0      1164
70000.0      1230
80000.0      1436
90000.0      1757
100000.0     2169
125000.0     5170
150000.0    32424
Name: odometer_km, dtype: int64

In [73]:
# remove outliers in price column
autos = autos[autos['price'].between(100, 5000000)]

In [74]:
# view sorted frequency count of unique values for the updated price column
autos['price'].value_counts().sort_index(ascending=False).head(10)

3890000.0    1
1300000.0    1
1234566.0    1
999999.0     2
999990.0     1
350000.0     1
345000.0     1
299000.0     1
295000.0     1
265000.0     1
Name: price, dtype: int64

# Remove Outliers

The code above removed rows from the data set where the price was greater than 5 million, or under 100. This prices are unrealistically high or low for a car. The values that remain in the data set should more realistically reflect the price for a car.

In [82]:
# understand the date distribution for 3 date columns
autos['last_seen'].value_counts().sort_index(ascending=False).head(20)

2016-04-07 14:58:50    3
2016-04-07 14:58:48    3
2016-04-07 14:58:46    1
2016-04-07 14:58:45    1
2016-04-07 14:58:44    3
2016-04-07 14:58:42    2
2016-04-07 14:58:41    1
2016-04-07 14:58:40    1
2016-04-07 14:58:38    1
2016-04-07 14:58:37    1
2016-04-07 14:58:36    3
2016-04-07 14:58:34    2
2016-04-07 14:58:33    2
2016-04-07 14:58:31    2
2016-04-07 14:58:29    3
2016-04-07 14:58:28    2
2016-04-07 14:58:27    2
2016-04-07 14:58:26    2
2016-04-07 14:58:25    1
2016-04-07 14:58:24    2
Name: last_seen, dtype: int64

In [85]:
# describe the date columns
autos['registration_year'].describe()

count    48230.000000
mean      2004.729795
std         87.892058
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

# Date statistics

The code above depicts that the 'registration_year' column contains a minimum value of 1000, which is unrealistic. The maximum value is 9999, which is also unrealistic. 

The following date ranges were observed for the respective date columns:

**date_crawled:** 
March 05 2016 to April 07 2016

**ad_created:**
June 11 2015 to April 07 2016

**last_seen:**
March 05 2016 to April 07 2016

In [90]:
# remove records with registration year below 1950 or above 2016
autos = autos[autos['registration_year'].between(1950, 2016)]

In [93]:
# view the registration years highest and lowest values and the frequency
autos['registration_year'].value_counts(normalize=True).sort_index(ascending=False).head(30)

2016    0.025938
2015    0.008200
2014    0.014285
2013    0.017285
2012    0.028226
2011    0.034915
2010    0.034246
2009    0.044906
2008    0.047690
2007    0.049049
2006    0.057595
2005    0.062817
2004    0.058242
2003    0.058113
2002    0.053452
2001    0.056732
2000    0.066982
1999    0.062148
1998    0.050495
1997    0.041540
1996    0.029240
1995    0.025744
1994    0.013509
1993    0.009063
1992    0.007941
1991    0.007294
1990    0.007164
1989    0.003690
1988    0.002870
1987    0.001554
Name: registration_year, dtype: float64

In [108]:
# aggregate the average price by unique brand

unique_brands = []

for brand in autos['brand']:
    if brand not in unique_brands:
        unique_brands.append(brand)
        
mean_brand_prices = {}

for record in unique_brands:
    selected_rows = autos[autos['brand'] == record]
    mean_brand_prices[record] = selected_rows['price'].mean()

In [112]:
# print the mean price by brand
for i in mean_brand_prices:
    print(i)
    print(mean_brand_prices[i])

toyota
5167.091062394604
land_rover
19108.091836734693
peugeot
3113.860549132948
ford
4074.959196290572
alfa_romeo
4100.915857605178
jaguar
11961.56338028169
sonstige_autos
24579.36842105263
dacia
5915.528455284553
honda
4119.109589041096
suzuki
4126.341818181818
renault
2496.940394314535
fiat
2836.8736310025274
mitsubishi
3439.10290237467
volkswagen
5636.200306122449
chevrolet
6759.885931558935
skoda
6409.609724047306
nissan
4756.659634317863
citroen
3796.26267281106
saab
3211.6493506493507
porsche
46788.444444444445
chrysler
3486.5766871165642
daihatsu
1649.655172413793
volvo
4993.208037825059
smart
3596.40273556231
hyundai
5411.075431034483
kia
6018.442073170731
subaru
4033.7551020408164
daewoo
1064.0579710144928
audi
9380.718547986076
mercedes_benz
8669.448637784726
lada
2688.296296296296
lancia
3444.877551020408
trabant
1846.5238095238096
opel
3005.9309720265646
seat
4433.419621749409
mazda
4129.774787535411
mini
10639.450980392157
bmw
8621.165459173682
jeep
11650.5
rover
1602.290

# Observations

The 'Rover' brand only costs 1602 on average, and sonstige_autos cost approximately 24,000 on average. Most car prices from the data set are between 3,000 and 9,000.

In [113]:
# calculate the the mean mileage and price for some of the top several brands
top_brands = ['audi', 'bmw', 'ford', 'volkswagen', 'opel', 'mercedes_benz']

avg_mileage_by_brand = {}
avg_price_by_brand = {}

for i in top_brands:
    sel_rows = autos[autos['brand'] == i]
    mn_price = sel_rows['price'].mean()
    mn_miles = sel_rows['odometer_km'].mean()
    avg_mileage_by_brand[i] = mn_miles
    avg_price_by_brand[i] = mn_price
    
mileage_series = pd.Series(avg_mileage_by_brand)
price_series = pd.Series(avg_price_by_brand)

mileage_price_by_brand = pd.DataFrame(mileage_series, columns=['mean_mileage'])
mileage_price_by_brand['mean_price'] = price_series

print(mileage_price_by_brand)

                mean_mileage   mean_price
audi           129245.400298  9380.718548
bmw            132723.712551  8621.165459
ford           124386.398764  4074.959196
mercedes_benz  131069.673962  8669.448638
opel           129405.312940  3005.930972
volkswagen     128807.142857  5636.200306


# Observation

There appears to be no significant correlation between average mileage and average price for some of the top brands. The most expensive of the set (Audi) has an average mileage somewhere in the middle of the set, whereas the second most expensive of the set (Mercedes Benz) has the highest mileage.